File: app-ui/src/lib/schemas/provider.ts
// Provider registry schemas, mirroring docs/SPEC-API/001-SPEC-API.md §7.4 and the table in
// docs/SPEC-UI/001-SPEC-UI.md §6.3.
//
// `category`, `routability`, and `kind` are plain strings rather than enums. The registry treats category
// as an open string, so a panel that enumerated it would fail to render a provider the API had just
// returned; the panel filters by the values it sees and shows the rest verbatim. §14 Q9 records the same
// decision for a gateway key's status.
//
// Two model routes are read from a provider's detail screen and they are not the same list. §7.6's
// catalog carries the filters the screen offers, so its shapes live in `model.ts`. §7.4's
// `/providers/{provider_id}/models` is the provider's own answer, which is the only list a custom node
// has: its models come from the base URL it points at rather than from the embedded registry. The shapes
// for that route are here, beside the provider they belong to.

import { z } from 'zod';
import { label, stringList } from './primitives';

// The pagination block SPEC-API §4 returns beside a paginated list.
const pageMeta = z.object({
	page: z.number().int(),
	per_page: z.number().int(),
	total: z.number().int()
});

// How the stored accounts for one provider are doing. A nested block rather than three sibling fields,
// because the list renders it as one thing.
export const schemaProviderStatusSummary = z.object({
	total: z.number().int().min(0),
	active: z.number().int().min(0),
	disabled: z.number().int().min(0),
	error: z.number().int().min(0),
	rate_limited: z.number().int().min(0)
});

export type ProviderStatusSummary = z.infer<typeof schemaProviderStatusSummary>;

export const schemaProvider = z.object({
	id: z.string().min(1),
	name: z.string().min(1),
	category: z.string().min(1),
	auth_type: z.string(),
	auth_modes: stringList,
	has_oauth: z.boolean(),
	no_auth: z.boolean(),
	routability: z.string(),
	endpoint_count: z.number().int().min(0),
	status_summary: schemaProviderStatusSummary
});

export type Provider = z.infer<typeof schemaProvider>;

export const schemaProviderList = z.object({
	data: z.array(schemaProvider),
	meta: pageMeta
});

export type ProviderList = z.infer<typeof schemaProviderList>;

// One non-chat service kind's endpoint and credential placement. `auth_header` is reported because the
// value `key` means a query parameter rather than a header.
export const schemaProviderMedia = z.object({
	kind: z.string().min(1),
	base_url: z.string(),
	auth_type: z.string(),
	auth_header: z.string(),
	format: z.string(),
	default_model: z.string(),
	model_count: z.number().int().min(0)
});

export type ProviderMedia = z.infer<typeof schemaProviderMedia>;

// The detail body embeds the list shape, so it repeats every list field plus the transport defaults. Those
// defaults are what decide reachability, which is why the screen shows them rather than the display name.
//
// `media` is declared because the API sends it and a shape the panel does not describe is a shape it
// cannot notice changing. It is not rendered on this screen: §6.3's header names five facts and media is
// not among them, and §6.8 owns the per-kind view.
export const schemaProviderDetail = schemaProvider.extend({
	base_url: z.string(),
	format: z.string(),
	url_suffix: z.string(),
	validate_url: z.string(),
	timeout_ms: z.number().int(),
	model_count: z.number().int().min(0),
	chat_model_count: z.number().int().min(0),
	media: z
		.array(schemaProviderMedia)
		.nullish()
		.transform((value) => value ?? []),
	deprecated: z.boolean(),
	deprecation_notice: z.string().optional(),
	website: z.string().optional()
});

export type ProviderDetail = z.infer<typeof schemaProviderDetail>;

// One model the provider's own `/models` answer carries (§7.4). A node's rows come from the upstream its
// base URL points at; a registry provider answers the rows the registry already declares.
//
// `name`, `kind`, `capabilities`, and `dimensions` are optional because the route's own row type omits
// them when it has nothing to say, and the panel reads only `id` and `name`. `source` and `warning` state
// where the list came from, which is not decoration: a node whose upstream could not be reached answers a
// fallback, and a client that could not tell that from a fresh answer would present a stale list as
// current. Both are optional because the gateway omits them when the list is the upstream's own
// (app-serv draft 017 F2 carries them), and the panel renders whichever it receives.
export const schemaProviderModel = z.object({
	id: z.string().min(1),
	name: z.string().optional(),
	kind: z.string().optional(),
	capabilities: stringList.optional(),
	dimensions: z.number().int().optional()
});

export type ProviderModel = z.infer<typeof schemaProviderModel>;

export const schemaProviderModelList = z.object({
	data: z.array(schemaProviderModel),
	source: z.string().optional(),
	warning: z.string().optional()
});

export type ProviderModelList = z.infer<typeof schemaProviderModelList>;

// The categories the reference registry measures. Offered as filter suggestions, not as a closed set: the
// list the API returns is authoritative, and a provider with a category outside this list still renders.
export const PROVIDER_CATEGORIES = ['apikey', 'oauth', 'free', 'media', 'local'] as const;

// The list query, matching exactly what the API reads: `category`, `routability`, and paging.
//
// §6.3 also asks for a search over name and ID. The API does not accept one: `ProviderListQuery` declares
// only category and routability, and the handler reads no other filter. §6.3's own pagination discipline
// forbids the alternative of pulling the registry into a client-side filter, and R-26 forbids shipping a
// search box that cannot search. So the control is absent until the API grows the parameter, which is
// recorded as a gap rather than worked around here.
export const schemaProviderQuery = z.strictObject({
	category: z.string().optional(),
	routability: z.enum(['native', 'connector']).optional()
});

export type ProviderQuery = z.infer<typeof schemaProviderQuery>;

// The one-line status a provider row shows. Written as a sentence rather than a set of numbers because the
// operator's question is whether anything needs attention, not what the exact counts are.
export function statusSummaryText(summary: ProviderStatusSummary): string {
	if (summary.total === 0) return 'No endpoint configured';
	if (summary.error > 0) return `${summary.error} failing`;
	if (summary.rate_limited > 0) return `${summary.rate_limited} rate limited`;
	if (summary.active > 0) return `${summary.active} active`;
	return `${summary.disabled} disabled`;
}

// The provider label used in a heading or a table cell. `label` is reused so a name is bounded the same way
// everywhere; the fallback keeps a minimal registry entry from rendering a blank row.
export const providerLabel = label;

File: app-serv/internal/schema/provider.go
// Package schema holds request/response DTOs and their validation rules.
//
// @file      internal/schema/provider.go
// @for       The provider registry read contracts (SPEC-API-001 §7.4).
// @uses      internal/registry.
// @reason    §7.4 serves the embedded registry over HTTP, and §8.1 requires the
//
//	panel to read a provider's routability before configuring an endpoint that
//	can never answer. The wire shape is fixed here rather than derived from the
//	YAML struct so a registry field rename cannot silently change the API.
//
// @author    Dodi Rusmana <rusmanadodi@kentangtech.com>
// @layer     schema
// @stability experimental
// @since     2026-09-18
package schema

import (
	"github.com/rusmanadodi2598/pannelAI/app-serv/internal/registry"
)

// ProviderListQuery is the decoded filter of GET /api/v1/providers. The
// category is validated against the measured set rather than a hardcoded one,
// because the reference treats category as an open string.
//
// Q is the free-text search over id and display name (PORT 002 D1). It is
// bounded here rather than in the service, because a limit on external input
// belongs with the DTO that describes it (AGENTS.md §1.4, §2.4).
type ProviderListQuery struct {
	Category    string `json:"category,omitempty"`
	Routability string `json:"routability,omitempty" validate:"omitempty,oneof=native connector"`
	Q           string `json:"q,omitempty" validate:"omitempty,max=120"`
}

// ProviderResponse is one provider row. StatusSummary reports how the stored
// accounts for this provider are doing, which is why it is a nested object and
// not three sibling fields: a client renders it as one block.
type ProviderResponse struct {
	ID            string                   `json:"id"`
	Name          string                   `json:"name"`
	Category      string                   `json:"category"`
	AuthType      string                   `json:"auth_type"`
	AuthModes     []string                 `json:"auth_modes"`
	HasOAuth      bool                     `json:"has_oauth"`
	NoAuth        bool                     `json:"no_auth"`
	Routability   string                   `json:"routability"`
	EndpointCount int64                    `json:"endpoint_count"`
	StatusSummary ProviderStatusSummaryDTO `json:"status_summary"`
}

// ProviderStatusSummaryDTO counts the stored endpoints of one provider by
// state. The authoritative per-endpoint state stays on the endpoint resource;
// this is a roll-up so the list screen does not have to fetch every account.
type ProviderStatusSummaryDTO struct {
	Total     int64 `json:"total"`
	Active    int64 `json:"active"`
	Disabled  int64 `json:"disabled"`
	Error     int64 `json:"error"`
	RateLimit int64 `json:"rate_limited"`
}

// ProviderList is the §7.4 list body. It carries a meta block because the
// registry is a large filtered set (94 entries in the P1 document), unlike the
// provider-node list which is hand-configured.
type ProviderList struct {
	Data []ProviderResponse `json:"data"`
	Meta Page               `json:"meta"`
}

// ProviderModelResponse is one model inside a provider. Kind is carried because
// a non-chat model is not routable through the chat data plane (§7.15), so a
// client that hid it would offer a model string that cannot answer.
type ProviderModelResponse struct {
	ID           string   `json:"id"`
	Name         string   `json:"name"`
	Kind         string   `json:"kind"`
	Capabilities []string `json:"capabilities"`
	Dimensions   int      `json:"dimensions,omitempty"`
}

// ProviderMediaResponse is one non-chat service kind's endpoint and credential
// placement (§7.4, §7.10). AuthHeader is reported because `key` means a query
// parameter, not a header.
type ProviderMediaResponse struct {
	Kind         string `json:"kind"`
	BaseURL      string `json:"base_url"`
	AuthType     string `json:"auth_type"`
	AuthHeader   string `json:"auth_header"`
	Format       string `json:"format"`
	DefaultModel string `json:"default_model"`
	ModelCount   int    `json:"model_count"`
}

// ProviderDetailResponse is the §7.4 detail body. It carries the transport
// defaults a client needs to reason about the provider plus the media blocks,
// because reachability is decided by format and credential placement rather
// than by the provider's display name.
type ProviderDetailResponse struct {
	ProviderResponse
	BaseURL         string                  `json:"base_url"`
	Format          string                  `json:"format"`
	URLSuffix       string                  `json:"url_suffix"`
	ValidateURL     string                  `json:"validate_url"`
	TimeoutMS       int                     `json:"timeout_ms"`
	ModelCount      int                     `json:"model_count"`
	ChatModelCount  int                     `json:"chat_model_count"`
	Media           []ProviderMediaResponse `json:"media"`
	Deprecated      bool                    `json:"deprecated"`
	DeprecationNote string                  `json:"deprecation_notice,omitempty"`
	Website         string                  `json:"website,omitempty"`
}

// ProviderModelList is the §7.4 model list body.
//
// Source states where the list came from, which is not decoration: a custom
// node's models are read from its own upstream, so a client that could not tell
// an upstream answer from a fallback would present a stale list as current. It
// replaces the constant `suggested` flag, which carried no information because
// every row answered true.
type ProviderModelList struct {
	Data []ProviderModelResponse `json:"data"`
	// Source is "upstream" or "registry" (service.ModelSource*).
	Source string `json:"source"`
	// Warning explains why Source is not upstream, in English. It is absent when
	// the list came from the upstream.
	Warning string `json:"warning,omitempty"`
}

// ProviderResponseFrom maps a registry entry and its stored account counts onto
// the list shape.
func ProviderResponseFrom(entry registry.Provider, summary ProviderStatusSummaryDTO) ProviderResponse {
	return ProviderResponse{
		ID:            entry.ID,
		Name:          providerDisplayName(entry),
		Category:      entry.Category,
		AuthType:      entry.AuthType,
		AuthModes:     append([]string(nil), entry.AuthModes...),
		HasOAuth:      entry.HasOAuth,
		NoAuth:        entry.NoAuth,
		Routability:   entry.ChatRoutability(),
		EndpointCount: summary.Total,
		StatusSummary: summary,
	}
}

// ProviderDetailFrom maps a registry entry onto the detail shape.
func ProviderDetailFrom(entry registry.Provider, summary ProviderStatusSummaryDTO) ProviderDetailResponse {
	chatModels := 0
	for _, model := range entry.Models {
		if model.IsChat() {
			chatModels++
		}
	}
	media := make([]ProviderMediaResponse, 0, len(entry.Media))
	for kind, config := range entry.Media {
		media = append(media, ProviderMediaResponse{
			Kind:         string(kind),
			BaseURL:      config.BaseURL,
			AuthType:     config.AuthType,
			AuthHeader:   config.AuthHeader,
			Format:       config.Format,
			DefaultModel: config.DefaultModel,
			ModelCount:   len(config.Models),
		})
	}
	sortMedia(media)
	return ProviderDetailResponse{
		ProviderResponse: ProviderResponseFrom(entry, summary),
		BaseURL:          entry.Transport.BaseURL,
		Format:           entry.Transport.Format,
		URLSuffix:        entry.Transport.URLSuffix,
		ValidateURL:      entry.Transport.ValidateURL,
		TimeoutMS:        entry.Transport.TimeoutMS,
		ModelCount:       len(entry.Models),
		ChatModelCount:   chatModels,
		Media:            media,
		Deprecated:       entry.Display.Deprecated,
		DeprecationNote:  entry.Display.DeprecationNotice,
		Website:          entry.Display.Website,
	}
}
